mlx/linalg.cpp

```diff
@@ -0,0 +1,64 @@
+// Copyright © 2023 Apple Inc.
+
+#include <iostream>
+#include <numeric>
+#include <set>
+#include <variant>
+
+#include "mlx/array.h"
+#include "mlx/linalg.h"
+#include "mlx/ops.h"
+
+namespace mlx::core::linalg {
+
+array vector_norm(
+    const array& a,
+    const std::variant<double, std::string>& ord,
+    const std::vector<int>& axes,
+    bool keepdims,
+    StreamOrDevice s) {
+  return std::visit(
+      overloaded{
+          [&](double p) {
+            if (p >= 1)
+              return power(
+                  sum(power(abs(a, s), array(p), s), axes, keepdims, s),
+                  array(1.0 / p),
+                  s);
+            else if (p == 0)
+              return sum(
+                  where(a != 0, array(1), array(0), s), axes, keepdims, s);
+            else
+              throw std::invalid_argument(
+                  "[core.linalg.norm] p norm is defined only for p >= 1.");
```

Comment:
Why not support `0 < p < 1`?

Reply:
https://statisticaloddsandends.wordpress.com/2020/05/27/lp-norm-is-not-a-norm-when-p-1/

Reply:
Yes it's technically not a norm, nevertheless it is supported by numpy and commonly used for sparsity. Since it's a trivial modification, I'd be in favor of adding it.

Reply:
I will support 0 < p < 1. :)